tests/dragonfly/utility.py

```diff
@@ -15,68 +20,413 @@ def grouper(n, iterable):
+        self.gen = CommandGenerator(
+            keys, val_size, batch_size, max_multikey
+        )
+        self.port = port
+        self.dbcount = dbcount
+        self.stop_flag = False
+
+    async def run(self, target_times=None, target_deviation=None):
+        """
+        Run a seeding cycle on all dbs either until stop(), a fixed number of batches (target_times)
+        or until reaching an allowed deviation from the target number of keys (target_deviation)
+        """
+        print(f"Running times:{target_times} deviation:{target_deviation}")
+        self.stop_flag = False
+        queues = [asyncio.Queue(maxsize=3) for _ in range(self.dbcount)]
+        producer = asyncio.create_task(self._generator_task(
+            queues, target_times=target_times, target_deviation=target_deviation))
+        consumers = [
+            asyncio.create_task(self._cosumer_task(i, queue))
+            for i, queue in enumerate(queues)
+        ]
+
+        await producer
+        for consumer in consumers:
+            await consumer
+
+    def stop(self):
+        """Stop all invocations to run"""
+        self.stop_flag = True
+
+    def reset(self):
+        """ Reset internal state. Needs to be called after flush or restart"""
+        self.gen.reset()
+
+    async def capture(self, port=None, target_db=0, keys=None):
+        """Create DataCapture for selected db"""
+        if port is None:
+            port = self.port
+
+        if keys is None:
+            keys = sorted(list(self.gen.keys_and_types()))
+
+        client = aioredis.Redis(port=port, db=target_db)
+        return DataCapture(await self._capture_entries(client, keys))
+
+    async def compare(self, initial_capture, port=6379):
+        """Compare data capture with all dbs of instance and return True if all dbs are correct"""
+        print(f"comparing capture to {port}")
+        keys = sorted(list(self.gen.keys_and_types()))
+        captures = await asyncio.gather(*(
+            self.capture(port=port, target_db=db, keys=keys) for db in range(self.dbcount)
+        ))
+        for db, capture in zip(range(self.dbcount), captures):
+            if not initial_capture.compare(capture):
+                eprint(f">>> Inconsistent data on port {port}, db {db}")
+                return False
+        return True
+
+    def target(self, key_cnt):
+        self.gen.key_cnt_target = key_cnt
+
+    async def _generator_task(self, queues, target_times=None, target_deviation=None):
+        cpu_time = 0
+        submitted = 0
+
+        while True:
+            if self.stop_flag:
+                break
+
+            start_time = time.time()
+            blob, deviation = self.gen.generate()
+            cpu_time += (time.time() - start_time)
+
+            await asyncio.gather(*(q.put(blob) for q in queues))
+            submitted += 1
+
+            if target_times is not None and submitted >= target_times:
```

Comment:
it seems that it would be even better to have a local function here that would monitor the exit condition that would use both target_deviation, target_times and stop_flag, it would make it more clear I think